hash/hash_comparator.go

```diff
@@ -60,24 +140,14 @@ func CompareArgon2id(_ context.Context, password []byte, hash []byte) error {
 	return ErrMismatchedHashAndPassword
 }
 
-func IsBcryptHash(hash []byte) bool {
-	res, _ := regexp.Match("^\\$2[abzy]?\\$", hash)
-	return res
-}
-
-func IsArgon2idHash(hash []byte) bool {
-	res, _ := regexp.Match("^\\$argon2id\\$", hash)
-	return res
-}
-
 func decodeArgon2idHash(encodedHash string) (p *config.Argon2, salt, hash []byte, err error) {
 	parts := strings.Split(encodedHash, "$")
-	if len(parts) != 6 {
+	if len(parts) != 5 {
```

Comment:
How is this not breaking previously stored hashes? I don't think that the indexes should change here, the format strings did not change anything regarding `$` positions.

Reply:
If we assume that after migration all `bcrypt` passwords will have prefix it should work. In this implementation `Compare()` will strip prefixes (via `ParsePasswordHash()`) and this function will receive hash without prefix `argon2id`, `bcrypt` etc. Take a look at the tests - high level `Compare()` receives the same input, implementation of the lower level `Compre*()` differs. I did it this way to have unified input to those function and avoid parsing/splitting hashes multiple time in the process and pass only relevant part of the hash down the call stack.